Update Iframe.js to use MutationObserver (#9300)

schnepe2 opened the pull request:
My proposal for #9300 using MutationObserver approach.
This **seems** to work, although the original poster should check if it's working for him/her

schnepe2 committed:
Update Iframe.js

oetiker commented on framework/source/class/qx/ui/embed/Iframe.js:
```diff
@@ -69,12 +69,25 @@ qx.Class.define("qx.ui.embed.Iframe",
+          return mutationRecord.addedNodes;
+        };
+
+        var observer = new MutationObserver(function (mutationRecords)
```
does the MutationObserver object exist on all gecko engine versions ?

schnepe2 replied:
...at least on those who complain about the Mutation Events ;)
According to [MDN](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver#Browser_compatibility) it's available for Firefox since Version 14.
Should we really add a environment-switch for that?

schnepe2 replied:
@oetiker : Independent of my question above, I think I will add a `qx.bom.client.EcmaScript#getMutationObserver` function.
Is that the right place for that information?

schnepe2 committed:
Update EcmaScript.js

schnepe2 committed:
Added getMutationObserver()

Added getMutationObserver() method
and "ecmascript.mutationobserver" environment variable

schnepe2 committed:
Added "ecmascript.mutationobserver" doc

schnepe2 committed:
Code wrapped in environment check

schnepe2 committed:
ignore(MutationObserver) added

schnepe2 committed:
@ignore(MutationObserver) added

level420 commented:
@schnepe2 current chrome is reported as "webkit" and also has `MutationObserver` implemented.

schnepe2 committed:
iframe out-of-sync handler applied for all engines

- iframe-out-of-sync handler applied for all engines
- removed qx.bom.Iframe.getWindow(iframeDomElement).stop() call, cause "edge" and "mshtml" prevent access to it and on all the other engines (I've tested) it seems not to be needed.
(Tested with: Edge, IE 11, FF 52, Chrome 57 & Palemoon 27.2.1)

schnepe2 commented:
@level420 yeah I've added the according environment check method and used it
4cd5ababb99092db2279653e8964660214979564 should be something to look at (sorry for the many single commits).

schnepe2 commented:
As asked by me in the issue discussion (#9300):
Can someone make a test case for this issue / work-around?

My test application is basically just this:
```
qx.Class.define("dummy.Application",
{
  extend : qx.application.Standalone,

  members :
  {
    // overridden
    main : function ()
    {
      this.base(arguments);

      var parent = new qx.ui.container.Composite(new qx.ui.layout.VBox());
      var container1 = new qx.ui.container.Composite(new qx.ui.layout.VBox());
      parent.add(container1, {flex:1});
      var container2 = new qx.ui.container.Composite(new qx.ui.layout.VBox());
      parent.add(container2, {flex:1});

      var iframe = new qx.ui.embed.Iframe();

      var doc = this.getRoot();
      doc.add(parent, {edge: 10});

      container1.add(iframe, {flex: 1});
      iframe.setSource("http://www.c64.com");

      window.setTimeout(function() {
        iframe.setSource("http://www.w3c.org");
        container2.add(iframe, {flex: 1});
      }, 5000);
    }
  }
});
```
But it can only be run and observed manually :-1: 
Writing a proper unit test for this would be needed, but that's something I have to assign to the professionals :grin:

schnepe2 committed:
(re-)added Window#stop() call

Due to the fact that I don't know under what circumstances the Window.stop() call might be necessary I've re added it again but this time protected by "isIE" checks (...whose Window object do not have a stop() method)

schnepe2 committed:
Issue 9300 tests (#2)

* Create hello.html

* Added test for issue #9300

Testcase for "SyncSourceAfterDOMMove" added

level420 commented on framework/source/class/qx/ui/embed/Iframe.js:
```diff
@@ -335,7 +380,11 @@ qx.Class.define("qx.ui.embed.Iframe",
+             qx.core.Environment.get("browser.name") != "ie" )
+        {
+          qx.bom.Iframe.getWindow(iframeDomElement).stop();
+        }
```
@schnepe2 why is this specific to ie and edge and was not before?
Ist this something specific to the mutation even?

schnepe2 replied:
Yes and No... the stop() method is not available in "Internet Explorer".
To **remove** the **Mutation Event** is all this issue is about!

level420 replied:
Ist this stop() needed for all browsers except ie/edge? Event if MutationObserver ist not present?

level420 replied:
And convinced on this point.

schnepe2 replied:
> Ist this stop() needed for all browsers except ie/edge?

To be honest, I don't think so.
But as I have no information about the initial need for this whole work-around I decided to keep it (as much as possible).

In my test(s) it makes no difference, but these test(s) only loaded small HTML pages ...
... I can imagine however, that if the first ("_pre-DOM-move_") URL loads lots of data it can make sense to stop loading that, when a "_DOM-move-and-URL-change_" was detected.
It might still be superfluous (as the URL is changed and the Client might stop loading the old URL anyway), but that's something I can neither confirm nor deny.

level420 commented on framework/source/class/qx/ui/embed/Iframe.js:
```diff
@@ -67,13 +68,57 @@ qx.Class.define("qx.ui.embed.Iframe",
+    }
+    else // !qx.core.Environment.get("ecmascript.mutationobserver")
+    {
+      this.addListenerOnce("appear", function ()
```
@schnepe2 before your PR the mutation event was attached to all gecko browsers. Now it is attached to every browser NOT having the mutation observer. 
Before the PR the listener was not attached to IE, Safari and Chrome which both report to be webkit.

I think adding this listener should be guarded with
```javascript
if ((qx.core.Environment.get("engine.name") == "gecko"))
```
Or am I wrong here?

schnepe2 replied:
Please ask the writer of the original work-around why this had been done only for gecko.
Possibly this work-around was only needed for gecko back then, but now it must be applied to all but Internet Explorer (both edge and IE).

level420 replied:
OK. Then you think that the else path should be applied to all browser NOT having MutationObserver?

schnepe2 replied:
I don't **think** that the else path should be applied to all browser not having MutationObserver, I have **tested** it (at least with some Browsers) AND applied additional Tests.

_Opinion-ON_
From my point of view this complete work-around should be removed
and anyone who thinks he/she must implement an application were it is needed to **move** an iframe from one area of the DOM into another should handle this "issue" in his/her application.
The original work-around supplier seemed to be important enough so that it was introduced into the framework.
_Opinion-OFF_

level420 replied:
OK convinced on this point.

oetiker reviewed (approved):
looks good

level420 commented:
2 approvals, more than 2 business days
merging
thank you @schnepe2
